plugins/stylelint-no-unknown-custom-properties/index.js

```diff
@@ -0,0 +1,148 @@
+/*!
+ * Copyright 2023 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+const path = require("path");
+const fs = require("fs");
+
+const stylelint = require("stylelint");
+
+const ruleName = "spectrum-tools/no-unknown-custom-properties";
+const messages = stylelint.utils.ruleMessages(ruleName, {
+    expected: (prop) => `Custom property '${prop}' not defined`,
+});
+
+const postcss = require("postcss");
+const valueParser = require("postcss-value-parser");
+
+/** @type {import('stylelint').Plugin} */
+const plugin = stylelint.createPlugin(ruleName, (enabled, { ignore = [], checkDependencies = false } = {}) => {
+    return (root, result) => {
+        const validOptions = stylelint.utils.validateOptions(
+            result,
+            ruleName,
+            {
+                actual: enabled,
+                possible: [Boolean, null],
+            },
+            {
+                actual: ignore,
+                possible: [String, RegExp],
+            },
+            {
+                actual: checkDependencies,
+                possible: [Boolean],
+            },
+        );
+
+        if (!validOptions) return;
+
+        /** @type {Set<string>} */
+        const localDefinitions = new Set();
+        const sharedDefinitions = new Set();
+
+        root.walkDecls(/^--/, ({ prop }) => {
+            localDefinitions.add(prop);
+        });
+
+        const sourceFile = root.source.input.file;
+        const parts = sourceFile ? sourceFile.split(path.sep) : [];
+        const rootIdx = parts.indexOf("components");
+        const componentRoot = parts.slice(0, rootIdx + 2).join(path.sep);
+
+        function fetchResolutions(depName) {
```

Comment:
Could you add a JSDoc style comment to this function, as well as some additional comments in this file for some of the larger blocks of code (e.g. line 73, 86, 107, etc)?

Reply:
I can but can we make that a post-merge to-do since it's not affecting functionality? Only ask because this is a step toward simplifying the build migration and the build migration is my main focus. I can add a backlog story to ensure this gets returned to.

Reply:
(Part of my hesitation to do it now is that I didn't write the plugin, it was written by Rajdeep for the build rework so to add a solid JS-doc, I'd need to read through the whole thing again and re-acquaint myself with it's inner workings)